cpp/include/cudf/io/detail/orc.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file orc.hpp
+ * @brief cuDF-IO reader classes API
+ */
+
+#pragma once
+
+#include <cudf/io/orc.hpp>
+
+namespace cudf {
+namespace io {
+namespace detail {
+namespace orc {
+/**
+ * @brief Class to read ORC dataset data into columns.
+ */
+class reader {
+ private:
+  class impl;
+  std::unique_ptr<impl> _impl;
+
+ public:
+  /**
+   * @brief Constructor from an array of file paths
+   *
+   * @param filepaths Paths to the files containing the input dataset
+   * @param options Settings for controlling reading behavior
+   * @param mr Device memory resource to use for device memory allocation
+   */
+  explicit reader(std::vector<std::string> const& filepaths,
```

Comment:
Eventually we'll want to remove file path based constructors completely and use `datasource` exclusively, but I doubt it's convenient for this PR.

Reply:
I would go with a follow-up PR, as it is mostly related to detail API rather than options.